controllers/volumereplicationgroup_controller.go

```diff
@@ -112,24 +111,63 @@ func pvcPredicateFunc() predicate.Funcs {
 
 			log.Info("Update event for PersistentVolumeClaim")
 
-			// TODO: If finalizers change then deep equal of spec fails to catch it, we may want more
-			// conditions here, compare finalizers and also status.state to catch bound PVCs
-			return !reflect.DeepEqual(oldPVC.Spec, newPVC.Spec)
+			return updateEventDecision(oldPVC, newPVC, log)
 		},
 		DeleteFunc: func(e event.DeleteEvent) bool {
-			log := ctrl.Log.WithName("pvcmap").WithName("VolumeReplicationGroup")
-
-			log.Info("delete event for PersistentVolumeClaim")
-
-			// PVC deletes are not of interest, as the added finalizers by VRG would be removed
-			// when VRGs are deleted, which would trigger the delete of the PVC.
+			// No need to reconcile VRG on pvc deletion.
+			// An update event with Nonzero deletion time
+			// would have been received to properly reconcile
+			// VRG.
 			return false
 		},
 	}
 
 	return pvcPredicate
 }
 
+func updateEventDecision(oldPVC *corev1.PersistentVolumeClaim,
+	newPVC *corev1.PersistentVolumeClaim,
+	log logr.Logger) bool {
+	const requeue bool = true
+
+	pvcNamespacedName := types.NamespacedName{Name: newPVC.Name, Namespace: newPVC.Namespace}
+	predicateLog := log.WithValues("pvc", pvcNamespacedName.String())
+	// If finalizers change then deep equal of spec fails to catch it, we may want more
+	// conditions here, compare finalizers and also status.phase to catch bound PVCs
+	if !reflect.DeepEqual(oldPVC.Spec, newPVC.Spec) {
+		predicateLog.Info("Spec changed. Requeing")
+
+		return requeue
+	}
+
+	if oldPVC.Status.Phase != corev1.ClaimBound && newPVC.Status.Phase == corev1.ClaimBound {
+		predicateLog.Info("PersistentVolumeClaim bound. ", "oldPVC Phase", oldPVC.Status.Phase,
+			"newPVC phase", newPVC.Status.Phase)
+
+		return requeue
+	}
+
+	// First check whether the pvc is being deleted or not. Then check for
+	// finalizer's presence. Because, if the pvc is being deleted and as part
+	// of handling it, finalizer might have been deleted in a previous reconcile
+	// loop. So, no need to add finalizer if a pvc is being deleted.
+	if !newPVC.GetDeletionTimestamp().IsZero() {
+		predicateLog.Info("Got delete event on pvc", newPVC.Name, newPVC.Namespace)
+
+		return requeue
+	}
+
+	// If VR protection finalizer is not there for the updated PVC, requeue.
+	if !containsString(newPVC.Finalizers, pvcVRFinalizerProtected) {
+		return requeue
+	}
+
```

Comment:
(generic comment based on other predicates)
The function `isPVCReadyForSecondary` should return a requeue status that the callers should skip (rather than requeue VRG), as the predicates will trigger the requeue when these conditions are met.

Also callers need to be able to distinguish between a requeue and a skip, which `isPVCReadyForSecondary` does not return as yet, and hence needs to change.

Reply:
Just to make sure I understand this correctly, are you suggesting that function "isPVCReadyForSecondary" should return 2 things (requeue and skip)? 

To ensure that when VRG finds that pvc is not ready for secondary, instead of explicitly requeing, it waits to receive notifications about changes, the only thing that needs to be changed is "returning !requeue" in "reconcileVRAsSecondary"  function.

So, does the function "isPVCReadyForSecondary" have to return "requeue and skip"? 

Let me know if I am missing something.

Reply:
> Just to make sure I understand this correctly, are you suggesting that function "isPVCReadyForSecondary" should return 2 things (requeue and skip)?

`isPVCReadyForSecondary` should return ready or !ready

> 
> To ensure that when VRG finds that pvc is not ready for secondary, instead of explicitly requeing, it waits to receive notifications about changes, the only thing that needs to be changed is "returning !requeue" in "reconcileVRAsSecondary" function.

`reconcileVRAsSecondary` should return requeue and skip, as it may get an error from `processVRAsSecondary` and hence requeue. If it gets !ready from `isPVCReadyForSecondary` it should skip.

> 
> So, does the function "isPVCReadyForSecondary" have to return "requeue and skip"?

No corrected above, original comment should have used `reconcileVRAsSecondary` in the second paragraph.

> 
> Let me know if I am missing something.